File: Sample/Scripts/Global/Gameplay/GameplayAbilitySpec.lua
local GameplayAbilitySpec = {}

GameplayAbilitySpec.ability = nil
GameplayAbilitySpec.sourceKey = nil
GameplayAbilitySpec.grantOrder = 0

function GameplayAbilitySpec:init(ability, sourceKey, grantOrder)
    self.ability = assert(ability, "Gameplay Ability is required")
    self.sourceKey = sourceKey
    self.grantOrder = grantOrder or 0
end

return class(GameplayAbilitySpec)

File: Game/Engine/Source/Global/src/Gameplay/GameplayAbilitySpec.cpp
#include <Gameplay/GameplayAbilitySpec.hpp>
#include <Gameplay/GameplayAbility.hpp>

#include <cmath>
#include <stdexcept>
#include <utility>

GameplayAbilitySpec::GameplayAbilitySpec(
    std::shared_ptr<GameplayAbility> gameplayAbility, RuntimeValue source,
    int order)
    : ability(std::move(gameplayAbility)),
      sourceKey(std::move(source)),
      grantOrder(order) {
    if (ability == nullptr) {
        throw std::invalid_argument("Gameplay Ability is required");
    }
}
